ENH: Initial implementation of Rotation.match_vectors

adbugger opened the pull request:
Solution of Wahba's problem using ~Davenport's q-method~ Markley's SVD method.

adbugger committed:
Initial commit for Rotation class

adbugger committed:
Modified docstring

adbugger committed:
Initial commit for overall package documentation

adbugger committed:
Corrections transform package main documentation and imports in __init__.py

adbugger committed:
Implemented from_quaternion. Class method not being documented.

adbugger committed:
Added tests for from_quaternion function

adbugger committed:
Shifted normalisation to `__init__` function

adbugger committed:
Added `normalized` flag. Class method still not documented.

adbugger committed:
Finally fixed documentation. All methods are now documented.

adbugger committed:
Reduced number of operations when normalizing. Added tests.

adbugger committed:
Added `transform/setup.py`, modified input shape check for clarity.

adbugger committed:
Minor change to if-else logic

adbugger committed:
STY: PEP8 compliance for `rotation.py` and `setup.py`

adbugger committed:
STY: Added actual argument shape in ValueError

adbugger committed:
MAINT: Zero norm quaternions now raise ValueError instead of UserWarning

adbugger committed:
STY: fixed whitespace in test_rotation.py

adbugger committed:
ENH: Added Rotation.as_quaternion method

The shape of the returned array depends on the initialization, and
whether the Rotation object contains a single quaternion.

adbugger committed:
TST: Tests for `Rotation.as_quaternion()` function.

Modified tests to work with the as_quaternion() function. Also
added a new test case to ensure that input of from_quaternion()
output of as_quaternion() always match in shape.

adbugger committed:
DOC: Added `as_quaternion` to methods section

adbugger committed:
ENH: Implemented as_dcm method.

adbugger committed:
TST: Added test cases for as_dcm

adbugger committed:
MAINT: Simpler and more efficient dcm component assignment

According to the tests, array indices during assignment are now
required to be transpose of the earlier intuition.

adbugger committed:
STY: Changed description of shape in docstrings

adbugger committed:
TST: Modify tests to use numpy.assert_equal instead of python assert

adbugger committed:
ENH: Added scalar implementation of from_dcm for testing only

adbugger committed:
MAINT: Updated experimental implementation

Forgot to git add to the last commit

adbugger committed:
ENH: Experimental implementation of vectorized approach

adbugger committed:
MAINT: Fixed array indexing

adbugger committed:
MAINT: Faster vector implementation.

The benchmarks on the wiki has been updated with the latest timings
and the vector implementation performs much better than the scalar
one. Moving on to cleaner implementation, documentation, and testing.

adbugger committed:
MAINT : Cleaned up code and decided on vector implementation.

adbugger committed:
TST: Added tests for from_dcm function

adbugger committed:
TST: Updated test to remove use of np.matmul

The matmul function was not introduced in numpy v1.8.2 and causes
travis-ci builds to fail. Instead of checking product of matrix and
transpose with identity, now we check equality of matrix inverse and
its transpose.

adbugger committed:
MAINT: normalized=True and np.empty in from_dcm

The `from_dcm` function now passes normalized quaternions to `__init__`
and uses np.empty to initialize `decision_matrix`.

adbugger committed:
TST: Fixed random state and random seed in tests

adbugger committed:
DOC: Updated documentation of from_dcm and added reference

adbugger committed:
TST: Used np.einsum instead of matmul

Travis-CI has been setup to test on a build with numpy==1.8.2, which
did not have np.matmul. To ensure test cases pass, Einstein summation
convention has been used to multiply matrices.

nmayorov committed:
MAINT: Normalize quat in a single place in Rotation.from_dcm

adbugger committed:
ENH: Implemented initialization from rotation vector

adbugger committed:
TST: Added test cases for `from_rotvec` function

adbugger committed:
STY: Fixed indentation for pep8

adbugger committed:
MAINT: Fixed implementation of from_rotvec

adbugger committed:
TST: Added identity quaternion in tests for `from_rotvec`

adbugger committed:
DOC: Added `from_rotvec` to documentation

adbugger committed:
MAINT: Fixed typo in function

adbugger committed:
MAINT: Renamed variable and used ** instead of np.power

adbugger committed:
MAINT: Changed implementation for clarity

Introduce `scale` variable in `from_rotvec` for clarity

adbugger committed:
TST: Added new tests and modified old ones for from_rotvec

Added a test case for small angles. Modified previous tests to use
as_quaternion function instead of accessing ._quat directly

adbugger committed:
ENH: Added implementation of `as_rotvec` function

adbugger committed:
STY: Removed redefinition of `small_norms` variable

adbugger committed:
STY: Style changes to code

adbugger committed:
TST: Better testing for small angles

adbugger committed:
STY: Removed extra blank line

adbugger committed:
TST: Shortened numbers in tests

adbugger committed:
MAINT: Used boolean masks instead of np.nonzero

adbugger committed:
TST: Added match parameter to pytest.raises(ValueError)

adbugger committed:
STY: Fixed whitespace

adbugger committed:
TST: Added test cases for `as_rotvec` function

adbugger committed:
ENH : Initial implementation of `from_euler`

Issues still need to be worked out in input format validation.
Implementation is still returning nans.

adbugger committed:
ENH: Implementation of `from_euler` function

Fixed errors and made a few decisions. Now, in the case of a single
rotations, the user can specify upto 3 angles and axes. In the case
of multiple rotations, however, numpy.asarray requires that the
dimensions match, so there the user must input a (N, 3) matrix only.

adbugger committed:
MAINT: Updated str-language specification

adbugger committed:
ENH: `from_euler` implementation using elementary quaternions

Introduced standalone functions for quaternion composition.

adbugger committed:
TST: Simple tests for `from_euler`

Initial tests. More to be added.

adbugger committed:
STY: Minor change to test file

adbugger committed:
TST: Changed test to compare quaternions. All tests pass now.

adbugger committed:
MAINT: Changed implementation of `from_euler`.

Implementation of `from_euler` has been changed according to
suggestions. Standalone functions, where appropriate, have been moved
out of the class.

adbugger committed:
TST: Add test case to check order of rotations

Intrinsic rotations should be same as extrinsic with the order of axes
reversed

adbugger committed:
MAINT: Minor changes to calculations and condition checks

adbugger committed:
TST: Elementary test to check extrinsic rotation implementation

adbugger committed:
MAINT: Changes to `from_euler`

adbugger committed:
TST: Add thorough testing for `from_euler`

adbugger committed:
STY/DOC: Rename constane, combine reference in docstring

adbugger committed:
ENH: Initial implementation of `as_euler` function. Still buggy.

adbugger committed:
MAINT: Fixed bugs with `as_euler`. Initial implementation now works.

Had missed and arccos function previously. Understood adaptations to
algorithm. Instead of reversing the sign of every angle, we need to
reverse the axis sequence and angle sequence for extrinsic rotations.

adbugger committed:
TST: Tests for `as_euler` function

adbugger committed:
DOC: Added references

The reference links are behind a paywall. I couldn't find a link to the
full texts.

adbugger committed:
STY: Style changes to code

Removed leading underscores from global function names. Renamed
argument. Calculated sin(lambda) and cos(lambda) explicitly. Removed
whitespace from docstring. Modified local variable in test.

adbugger committed:
STY: Explanation in comment is clearer

adbugger committed:
TST: Test cases for `as_euler` with extrinsic rotations

These test cases compare the actual angles returned by the algorithm. It appears that we cannot always guarantee that these will be th same as the input angles, only that the final rotation generated by the angles will be the same.

For intrinsic rotations, it appears that the algorithm never gives the same angles.

adbugger committed:
MAINT: Used np.einsum instead of loop

adbugger committed:
TST: Changed variable names in test

adbugger committed:
MAINT/DOC: Corrected einsum broadcasting. Added function to docs.

adbugger committed:
DOC: Removed redundant reference. Cleared up convention difference.

adbugger committed:
MAINT: Enforced angle restrictions

adbugger committed:
MAINT: Updated range restrictions for second angle

adbugger committed:
TST: Use randomly generated angles for testing.

Now that the ranges for angle2 have been properly fixed, we can test for
angle equality with randomly generated angles within those ranges.

adbugger committed:
TST: Split tests according to axes type, symmetric and asymmetric

adbugger committed:
MAINT: Renamed variables appropriately

adbugger committed:
DOC: Clarified angle ranges in the docstring

adbugger committed:
STY/DOC: Style changes to code, fixed error in docstring

adbugger committed:
TST: Add tests to ensure proper handling of gimbal lock

adbugger committed:
MAINT: Fixed detection of gimbal lock

The problem was that the value of eps had been taken to be too small. It
is now the smallest value that passes the tests. (~1e-7). Also make some
minor style and doc changes.

adbugger committed:
DOC: Mentioned behaviour for cases of gimbal lock in docstring

adbugger committed:
TST: Tests for gimbal lock cases

adbugger committed:
MAINT: Don't apply angle adjustment for gimbal locked cases

adbugger committed:
TST: Gimbal lock can only be caused by second angle being close to 90

adbugger committed:
DOC: Removed explanation, added reference for gimbal lock

adbugger committed:
STY: Minor changes to code style

adbugger committed:
MAINT: Handling Gimbal lock.

Clarified warning for gimbal lock. Also tried not to reverse angle
sequences for gimbal locked cases but that leads to test failures.

adbugger committed:
MAINT: Handle extrinsic rotations in a separate branch

This is to ensure that after reversing angle sequence we still get third
angle as zero

adbugger committed:
TST: Angles specific to gimbal lock cases

adbugger committed:
MAINT: Remove dcm transposition and adapt rest of algorithm

adbugger committed:
DOC: Fixed docstring

adbugger committed:
MAINT: Add leading '_' to private functions

adbugger committed:
ENH: Implementation of `inv` function

adbugger committed:
TST: Test for correct inversion of rotation

adbugger committed:
MAINT: `inv()` now respects `_single` flag

adbugger committed:
TST: Tests for single rotation inversion

adbugger committed:
MAINT: Fixed error with `_single` flag

adbugger committed:
MAINT: Concise implementation

adbugger committed:
Merge pull request #13 from adbugger/rotation_inv

Initial implementation of the inverse function

adbugger committed:
ENH: Initial implementation of rotation composition

adbugger committed:
MAINT: Simpler implementation

adbugger committed:
MAINT: Removed unused function

Since we're not using the matrix representation anymore, remove the
function responsible for generating that representation.

adbugger committed:
MAINT: Added error message for shape mismatch.

adbugger committed:
Merge pull request #14 from adbugger/rotation_mul

ENH: Initial implementation of rotation composition

adbugger committed:
DOC: Documentation changes for rotation composition function

adbugger committed:
DOC: Add __mul__ to exclusion list in class.rst

adbugger committed:
ENH: Initial implementation of `apply` function

adbugger committed:
MAINT/DOC: Explain rotation-projection duality, add inverse flag

adbugger committed:
DOC: Better explanation of rotation-projection duality

adbugger committed:
ENH: dot semantics for apply

adbugger committed:
DOC: Shorter explanation of output cases

adbugger committed:
TST: Initial tests for rotation application

adbugger committed:
ENH: Implementation of __getitem__ function

adbugger committed:
DOC: Add __getitem__ to exclusion list

adbugger committed:
ENH: Add copy functionality to __init__

adbugger committed:
ENH: from_quaternion now uses copy flag

adbugger committed:
ENH: from_dcm uses copy flag

adbugger committed:
ENH: from rotvec uses copy flag

adbugger committed:
ENH: from_euler uses copy flag

adbugger committed:
ENH: inv uses copy flag

adbugger committed:
ENH: __mul__ uses copy flag

adbugger committed:
ENH: Add number of rotation as a property without setter

adbugger committed:
MAINT: Apply function uses num_rotations property

adbugger committed:
DOC: Small changes to docstrings.

adbugger committed:
TST: Tests use np.expand_dims function

adbugger committed:
TST: Add test with inverse=True

adbugger committed:
DOC/STY: Minor docstring and style changes

adbugger committed:
TST: Add tests for indexing and get item

adbugger committed:
Merge pull request #15 from adbugger/rotation_apply

ENH: Initial implementation of `apply` function

adbugger committed:
TST: Update previous tests to resolve merge conflicts

adbugger committed:
TST: Update previous tests to resolve merge conflicts

adbugger committed:
Merge branch 'rotation' into rotation_index_copy

adbugger committed:
Merge branch 'rotation_index_copy' of https://github.com/adbugger/scipy into rotation_index_copy

adbugger committed:
MAINT: Remove copy flag from `from_quaternion`

adbugger committed:
Merge pull request #17 from adbugger/rotation_index_copy

ENH: Indexing, copying and addition of number property to rotation class

adbugger committed:
MAINT: Explicit imports in `__init__.py`

adbugger committed:
MAINT: as_quaternion returns a copy of quaternion array

adbugger committed:
MAINT: Replace property `n` with `len` function

adbugger committed:
TST: Test to check against accidental corruption of object

adbugger committed:
DOC: Rotation class docstring changes.

Listed representations and operations on rotaitons.

adbugger committed:
DOC: Changes to from_quaternion docstring

Added a returns section along with reference to quaternions for spatial
rotation.

adbugger committed:
DOC: Add returns section and references to as_quaternion

adbugger committed:
DOC: Returns section, docsting imrovements for ..._dcm functions

adbugger committed:
DOC: Remove reference to class internals

adbugger committed:
DOC: Returns section and docstring improvements for rotvec methods

adbugger committed:
DOC: Better documentation for euler methods

adbugger committed:
DOC: Documentation for inverse and composition

adbugger committed:
DOC: Improved documentation for rotation application

adbugger committed:
DOC: Documenting rotation indexing

adbugger committed:
DOC: Documentation for len function

adbugger committed:
DOC: Reorder method list. Remove section from package docstring.

adbugger committed:
STY: Reorder imports, remove unhelpful comments

adbugger committed:
STY: Reorder functions in code. Rename ..._quaternion to ..._quat.

adbugger committed:
DOC: Changed function docstrings.

More specific output names. Added references for rotvecs and dcms.
Simplified two-to-one mapping explanation in as_quat.

adbugger committed:
BUG: Fix nan error in degenerate euler axes

For certain gimbal locked cases, it appears that certain entries in
`dcm_transformed` are greater than unit norm. That results in nan values
when arccos function us used.

adbugger committed:
STY: Style changes to code

adbugger committed:
DOC: Documentation improvements

adbugger committed:
STY: Use len instead of shape[0]

adbugger committed:
ENH: Initial implementation of Rotation.estimate

Solves Wahba's problem using Davenport's q-method.

adbugger committed:
TST: Simple test to check functionality

adbugger committed:
STY: Whitespace changes

adbugger committed:
DOC: Add estimate to methods list

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def estimate(cls, vectors, transformed_vectors, coeff=None):
```
There are potentially many ways to estimate a rotation, I suggest to name it as `match_vectors` perhaps.

Definitely not `coef`, name it `weights`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
+
+        Parameters
+        ----------
+        vectors : array_like, shape (N, 3) or (3,)
```
You can't estimate rotation using a single vector.

Also remember we discussed what to do with vector norms. I think we should introduce `normalized=False` for that.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
+        vectors : array_like, shape (N, 3) or (3,)
+            Vectors expressed in initial frame A.
+        transformed_vectors : array_like, shape (N, 3) or (3,)
+            Result of applying an arbitrary rotation to `vectors`. Shape must
```
Use the similar description.

Also you can/should emphasize that two sets of vectors are typically measured somehow in the two frames. For example, directions to certain stars are known from datum in the global frame and they are measured (observed) in spacecraft frame, but with some angular errors. This way the attitude of spacecraft is determined. This is how conceptually this algorithm is used.

So your description is conceptually misleading, while formally sort of correct.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
+
+        Given `vectors` expressed in an initial frame A, corresponding
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
```
So the estimated DCM C does `v_A = C v_B`?

Also explain that it is estimated in (weighted) least-squares sense, writing objective function would be appropriate.

adbugger replied:
No, I meant `v_B = C v_A`. I will clear this confusion in the next push.

nmayorov replied:
I think I would prefer having `v_A, v_B` and `C^A_B`, such that `v_A = C^A_B v_B`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
+
+        The rotation is estimated using Davenport's q-method [1]_.
```
I suggest to use Markley SVD method, it is cleaner to implement and additionally conveniently provides the covariance matrix for the estimation.

And I suggest to return the covariance to be slightly more advanced and it is useful sometimes. Considering that, I think you should make a standalone function like `match_ vectors`.

What do you think of this suggestion?

adbugger replied:
When you say covariance matrix, do you mean the singular value decomposition of `A_opt` or the error covariance matrix? [Paper](https://pdfs.semanticscholar.org/f151/97510b1e501876a6f6d8683ac69ab1ef8d39.pdf) for reference.

I'm not exactly sure about the need for a standalone function, but the Markley SVD method does seem cleaner and more informative.

nmayorov replied:
I mean covariance of attitude error as a small rotation vector. 

Well, I think you are right that we don't need a separate function.

nmayorov replied:
To clarify, I think you don't need to consider the scale of the covariance only its "structure".

Check the formula (52) in the paper I sent you, what I suggest to return is  the matrix part withou "lambda_0 sigma_tot^2". At the moment though I forgot about fine points of this derivation, you need to figure it out. My idea is that we should return something like "sensitivity matrix" and the covariance can be computed by scaling it appropriately.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -778,3 +778,11 @@ def test_quat_ownership():
     assert_allclose(s._quat[0], np.array([1, 0, 0, 0]))
+
+
+def test_estimate_no_rotation():
```
Add a test with non-trivial rotation, two vectors and many vectors, without noise and with noise. Obviously with noise you won't get the rotation back precisely, but you can control the level of error, especially if you compute the covariance matrix.

nmayorov reviewed (commented):
Made some comments (from the phone).

nmayorov commented:
I mean small vector errors covariance.

Well, yes I believe standalone function is not necessary.
Sent using Zoho Mail


---- On Sun, 24 Jun 2018 22:14:03 +0500 notifications@github.com wrote ----

@adbugger commented on this pull request.

In scipy/spatial/transform/rotation.py:

> @@ -866,3 +868,94 @@ def __getitem__(self, indexer):
                   array.
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def estimate(cls, vectors, transformed_vectors, coeff=None):
+        # TODO: Better description of coeff?
+        """Estimate the rotation that best transforms a set of vectors to
+        another.
+
+        Given `vectors` expressed in an initial frame A, corresponding
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
+
+        The rotation is estimated using Davenport's q-method [1]_.
When you say covariance matrix, do you mean the singular value decomposition of A_opt or the error covariance matrix? Paper for reference.

I'm not exactly sure about the need for a standalone function, but the Markley SVD method does seem cleaner and more informative.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub, or mute the thread.

adbugger committed:
MAINT: Changes to rotation estimation

- Renamed `estimate` to `match_vectors.`
- Explained role of `weights` using loss function.
- Used SVD method instead of q-method.
- Return estimated rotation and sensitivity matrix

adbugger commented:
I have updated the implementation. Now we return an appropriate 'sensitivity_matrix`, along with the estimated rotation.

I need some help with the testing part here. I know I can create test cases by starting with fixed `vectors` and a given rotation, applying the rotation and adding [gaussian?] noise to get `transformed_vectors`. How do I ensure that the estimated rotation is within the error bounds? For example, if we were returning the full covariance matrix, I could possibly compute the error vector between `estimated_rotation.apply(vectors)` and `transformed_vectors`, and check that the result was within the tolerance given by the covariance matrix. Now since our `sensitivity_matrix` is scaled, I'm not sure how to proceed.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+            L(A) = \\frac{1}{2} \\sum_{i = 1}^{n} a_i \\lVert \\mathbf{b}_i -
+            A \\mathbf{r}_i \\rVert^2 ,
+
+        where :math:`a_i`'s are the `weights` corresponding to each vector, and
```
w_i for weights?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+
+    @classmethod
+    def match_vectors(cls, vectors, transformed_vectors, weights=None):
+        # TODO: Better description of coeff?
```
Remove.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+    @classmethod
+    def match_vectors(cls, vectors, transformed_vectors, weights=None):
+        # TODO: Better description of coeff?
+        """Estimate the rotation that best transforms a set of vectors to
```
Perhaps: "Estimate the rotation to match two set of vectors".

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+            Contains a single rotation which is the best estimate of the
+            rotation that transforms `vectors` to `transformed_vectors`.
+        sensitivity_matrix : `numpy.ndarray`, shape (3, 3)
+            Scaled covariance matrix of the three component error vector
```
This definitely needs to be explained better. Scaled in what way, how a user should compute a true covariance. Perhaps we should return the true covariance?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+        """Estimate the rotation that best transforms a set of vectors to
+        another.
+
+        Given `vectors` expressed in an initial frame A, corresponding
```
I still think that `transformed_vectors` is not a great name.

I don't know, maybe use something neutral like `a` and `b`: ``Find a rotation between frames A and B which best matches a set of unit vectors `a` and `b` observed in these frames. The following optimization problem is solved for the direction cosine matrix `C`: ... ``

My preference is `a = C b`, i.e. C projects from B to A or rotates frame A into B.

Not sure though.

adbugger replied:
My thought process behind this name was that it is immediately clear which set of vectors the `estimated_rotation` is being applied to.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+            Coefficients describing the relative importance of the vectors in
+            `vectors`. Number of values in `weights` must match number of
+            vectors specified in `vectors` and `transformed_vectors`. If None,
+            then all values in `weights` are assumed equal to +1.
```
I think "assumed to be equal" is enough.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+        Returns
+        -------
+        estimated_rotation : `Rotation` instance
+            Contains a single rotation which is the best estimate of the
```
Best estimate of the rotation to match `a` and `b`. Or something like that.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+                                 "{} values and {} vectors.".format(
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
```
I think this is a slight overuse of `einsum`. There are `np.outer` and other stuff, which might be more efficient. I don't know, perhaps nothing is wrong in using `einsum`, but I somehow don't like it. Judge yourself.

adbugger replied:
I looked at `np.outer` and that performs outer product of 1D vectors. When given an array of vectors like in our application, it performs outer products of every possible pair.
We need to perform outer products of corresponding vectors and sum the resultant matrices. I think this is a succinct way to do it. I may be wrong here; I'll change it if I find a better way.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+                             "covariance. It is impossible to determine the "
+                             "rotation uniquely.")
+
+        return cls.from_dcm(np.einsum('ij,jk->ik', u, vh)), p
```
This is just strange, why do use prefer `einsum` over a simple `dot`?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+        k = s[0]*s[1] + s[1]*s[2] + s[2]*s[0]
+        psi = (s[0]+s[1]) * (s[1]+s[2]) * (s[2]+s[0])
+
+        p = np.diag((k, k, k)) + np.einsum('ij,kj->ik', B, B)
```
`k * np.eye(3)`

And another very strange `einsum`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
+        u, s, vh = np.linalg.svd(B)
+        k = s[0]*s[1] + s[1]*s[2] + s[2]*s[0]
+        psi = (s[0]+s[1]) * (s[1]+s[2]) * (s[2]+s[0])
```
I believe `zeta` is used in the paper.

And for `k`, I guess `kappa`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
+        u, s, vh = np.linalg.svd(B)
```
I suggest to compute `A` right here to make it explicit. Now it is confusing.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+        if vectors.ndim != 2 or vectors.shape[-1] != 3:
+            raise ValueError("Expected input `vectors` to have shape (N, 3)"
+                             ", got {}.".format(vectors.shape))
+        if vectors.shape[0] == 1:
```
I guess check shape equality here. The check that the number of vectors > 1.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
+                             "vector.")
+
+        outvecs = np.asarray(transformed_vectors)
+        if outvecs.ndim != 2 or outvecs.shape[-1] != 3:
```
Doesn't seem necessary.

nmayorov reviewed (commented):
Reviewed the latest changes.

nmayorov commented:
> I need some help with the testing part here. I know I can create test cases by starting with fixed vectors and a given rotation, applying the rotation and adding [gaussian?] noise to get transformed_vectors. How do I ensure that the estimated rotation is within the error bounds? For example, if we were returning the full covariance matrix, I could possibly compute the error vector between estimated_rotation.apply(vectors) and transformed_vectors, and check that the result was within the tolerance given by the covariance matrix. Now since our sensitivity_matrix is scaled, I'm not sure how to proceed.

You need to figure out how the covariance is computed, it's probably a harder part of the paper. Note that there is an appendix. It's quite clear that the covariance is well defined and you can compute it from the noise you are using.

I believe you should add a small gaussian noise and then normalize the vectors. By the way, implement this normalization logic inside the function.

adbugger committed:
MAINT: Changes to Rotation.match_vectors

- Add normalized=False flag
- Change order of application on vectors
- Improve documentation about loss function
- Remove extra use of np.einsum
- Add test case without noise

adbugger commented:
Made some changes to the function. The next task is to figure out the covariance matrix.

adbugger committed:
MAINT: Normalize weights and return covariance matrix.

adbugger committed:
TST: Rotation estimation with noisy vectors.

adbugger commented:
Added a test case based on my understanding of the covariance matrix. Now that we're normalizing the weights to sum to unity; our previous `sensitivity_matrix` equals the full `covariance_matrix`. 

Cannot rebase as of now due to merge conflicts. I will use the editor to fix those conflicts once we have finalised the `rotation` branch.
